src/utils/BucketUtils.ts

```diff
@@ -0,0 +1,384 @@
+    })
+    term.dataUse?.secondary?.forEach((dut: DataUseTerm) => {
+      // Set the type for secondary data use terms to permissions
+      dut.type = ControlledAccessType.modifiers
+    })
+  })
+  // Create a map of DataUse to a list of datasets. This will serve as the basis for bucketing datasets by DataUse.
+  // Note that we need to use a string value of the DataUse object to ensure that we can use it as a key in a Map.
+  const datasetTermMap: Map<string, Dataset[]> = new Map<string, Dataset[]>()
+  terms?.forEach((term: DatasetTerm) => {
+    const stringValue = JSON.stringify(term.dataUse)
+    if (datasetTermMap.has(stringValue)) {
+      datasetTermMap.get(stringValue)?.push(datasets.filter((dataset: Dataset) => dataset.datasetId === term.datasetId)[0])
+    }
+    else {
+      datasetTermMap.set(stringValue, [datasets.filter((dataset: Dataset) => dataset.datasetId === term.datasetId)[0]])
+    }
+  })
+
+  // Iterate through the datasetTermMap to create buckets
+  const iterator = datasetTermMap.keys()
+  for (const key of iterator) {
+    const dataUseSummary: DataUseSummary = isNil(key) ? undefined : JSON.parse(key)
+    const datasets: Dataset[] = datasetTermMap.get(key) || []
+    const datasetIds = datasets?.map((dataset: Dataset) => dataset.datasetId)
+    const dacIds = datasets?.map((dataset: Dataset) => dataset.dacId) || []
+    const dacs: DacTerm[] = terms
+      .filter((term: DatasetTerm) => dacIds.includes(term.dacId))
+      .map((term: DatasetTerm) => term.dac)
+
+    const bucket: Bucket = {
+      key: '',
+      label: '',
+      datasets: datasets,
+      datasetIds: datasetIds,
+      dataUse: dataUseSummary,
+      dataUses: [...dataUseSummary?.primary || [], ...dataUseSummary?.secondary || []],
+      elections: [],
+      votes: [],
+      matchResults: [],
+      dacs: dacs,
+    }
+    buckets.push(bucket)
+  }
+
+  // The following steps are all bucket-centric, so we can process them in a single loop
+  // Steps 2-6
+  buckets.forEach((b: Bucket) => {
+    // Step 2: Find match results for each dataset in bucket
+    matchData.forEach((m: MatchResult) => {
+      b.datasets.forEach((dataset: Dataset) => {
+        if (dataset.datasetIdentifier.toLowerCase() === m.consent.toLowerCase()) {
+          b.matchResults.push(m)
+        }
+      })
+    })
+
+    // Step 3: Populate elections for datasets in this bucket
+    b.elections = findElectionsForDatasets(recentDar, b.datasetIds)
+
+    // Step 4: Populate votes for each bucket
+    b.votes.push(processVotesForBucket(b.elections))
+
+    // Step 5: Generate bucket key and label
+    b.key = 'bucket-' + b.datasetIds.join('-')
+    b.label = b.dataUses?.map((du: DataUseTerm) => du.code).join(', ') || 'Undefined Data Use'
+
+    // Step 6: Coalesce match results into a single result per bucket
+    b.algorithmResult = calculateAlgorithmResultForBucket(b)
+  })
+
+  // Step 7: Populate RUS Vote bucket with RP votes and move to the top of the bucket array
```

Comment:
I like how this function outlines all the steps in the docstring, then again piecewise near the relevant implementation.  Nicely maintainable handling of domain-specific complexity.